### ISSUE TITLE
javascript.py2js() output is not usable

### ISSUE THREAD

**Issue by benjie-git** [brython-dev/brython#2080]
I have a page that loads quite a bit of python in the main page, and on a worker thread.  Both need to be up and running before the page can display properly.  Right now, this takes a few seconds to load brython and the stdlib, load all of my python modules and dependent modules and convert them to js, and then run the js.

What is the best practice to speed up load time?  Is there a way to pre-compile all necessary python modules to one js file per thread?  Running javascript.py2js() on all of the python code per thread, and then trying to run its output is giving me errors (see below).

My page also needs to load the full brython and stdlib to allow running exec() on additional python code after loading, but I'm imagining that I could speed things up if brython didn't need to convert all of my python to js before running it, on every page load.

Thanks!

**Comment by benjie-git** [comment]
I've tried using javascript.py2js() on my python code to create a js version, which seems to work, but I haven't been able to use the js version as a replacement for my python code.  I get runtime errors about 'locals' not being defined, and so locals.__name__ can't be set, right near the beginning.

**Comment by benjie-git** [comment]
I'm taking the output from py2js() of my complete python code, and using that as a .js file that I load in a <script> tag. The first lines are:

    // Javascript code generated from ast
    var $B = __BRYTHON__,
        _b_ = $B.builtins,
        locals_exec = $B.imported["exec"],
        locals = locals_exec,
        frame = ["exec", locals, "exec", locals]
    frame.__file__ = '<string>'
    locals.__name__ = 'exec'


and it looks like `__BRYTHON__.imported["exec"]` is not defined, which is causing the problem.

**Comment by PierreQuentel** [comment]
> What is the best practice to speed up load time?

It is explained in the [documentation](https://www.brython.info/static_doc/en/import.html) : use `brython-cli modules`, this will group the Python source code of all the modules / packages used by your application in a single JS script. The indexedDB cache will store the Javascript translation of these modules; the translation is only done once at each new version of the JS script, not at each page load. The first run might take some time but the next ones will be much faster.

If you tried this approach and still had long loading times, can you give an example ?

> Is this usage not supported?

It is indeed not supported the way you expect. The result of `javascript.py2js()` is not meant to be executed as a standalone Javascript program; the Brython engine sets an environment to make it work (`__BRYTHON__.imported["exec"]` , as you guessed, but also values necessary if your code has imports)

**Comment by benjie-git** [comment]
Ahhh, thank you.  I had misunderstood this section of the documentation.  I thought `brython-cli make_modules` just removed unused modules from stdlib.  I didn't realize it also added our application modules as well!

One followup question:  My application allows running further python code from strings after compile time, using exec(), and I don't actually want to limit the stdlib modules available to that code.  Is there a way to add my custom application modules to brython_modules.js, without removing the currently unused stdlib modules?

Thanks very much!

**Comment by benjie-git** [comment]
I think I found my answer, that no, this is not currently supported by `brython-cli make_modules`.  I copied the make_modules code into my own little script, and modified it to do what I need.  Would you accept a pull request to add a similar `finder.modules.update(stdlib)` functionality as an option to `brython-cli make_modules`?  Maybe something like `brython-cli make_modules --keep-full-stdlib`?

    import os
    from brython import list_modules
    
    print('searching brython_stdlib.js...')
    stdlib_dir, stdlib = list_modules.load_stdlib_sitepackages()
    
    print('finding packages...')
    user_modules = list_modules.load_user_modules()
    finder = list_modules.ModulesFinder(stdlib=stdlib, user_modules=user_modules)
    finder.inspect()
    
    # Add the full stdlib back into the carefully optimized set of necessary modules
    # Not the most efficient solution, but a clean change
    finder.modules.update(stdlib)
    
    output_path = os.path.join(stdlib_dir, "brython_modules.js")
    finder.make_brython_modules(output_path)

**Comment by PierreQuentel** [comment]
In this case, there is another option which is already available: you can put all the logic of your application in a separate folder and generate a [Brython package](http://localhost:8000/static_doc/en/brython-packages.html) for it.

Then you include both **`brython_stdlib.js`** and **`my_brython_package.brython.js`** in the HTML page. Brython packages are handled by the indexedDB cache.

**Comment by benjie-git** [comment]
Thank you again.  I tried the make_package process, and I noticed that I need to manually call importScripts() for my custom_app.brython.js file in my worker thread, whereas my brython_modules.js is automatically imported into the worker thread.  This would be fine, but app startup was noticeably slower with this setup, so I'm back to using my custom make_modules script that pulls the full stdlib back into the generated modules file.

If you think that brython packages loaded into the main thread should also be automatically loaded into the worker thread, as is done with the stdlib, then we could leave this issue open.  Otherwise, I'm happy to have it closed.
